util/design/gen-top-docs.py

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+r"""Generates top level documentation from an hjson configuration file."""
+
+import argparse
+import logging as log
+from pathlib import Path
+
+import hjson
+from lib import common
+from tabulate import tabulate
+
+
+def to_markdown(text):
+    """Adds prefix and sufix newlines to 'text'."""
+    doc_text = f"""
+{text}
+"""
+    return doc_text
+
+
+def generate_mmap_table(top_level):
+    """Generates top level memory map table."""
+    header = ["Name", "Type", "Byte Address"]
+    table = [header]
+    colalign = ("left", ) * len(header)
+
+    for module in top_level["module"]:
+        for j, (name, base) in enumerate(module["base_addrs"].items()):
+
+            base_address = f"{base} ({name})"
+            if name == "null":
+                base_address = f"{base} (regs)"
+
+            if j == 0:
+                row = [module["name"], module["type"], base_address]
+            else:
+                row = ["", "", base_address]
+
+            table.append(row)
+
+    return tabulate(table,
+                    headers="firstrow",
+                    tablefmt="pipe",
+                    colalign=colalign)
+
+
+def generate_pinout_table(top_level):
```

Comment:
Yeah I do something very similar already in this file here:
https://github.com/lowRISC/opentitan/blob/master/util/topgen/gen_top_docs.py

It is being run as part of topgen and it currently produces the pinmux tables here:
https://docs.opentitan.org/hw/ip/pinmux/doc/#pinout-and-pinmux-mapping

These tables are more detailed, since they also print out the names for the pinmux insel/outsel constants (e.g. `kTopEarlgreyPinmuxInselIob9`), so it probably makes sense to keep both table versions (the one in the pinmux docs, and the one here in the toplevel spec).

The tables are however generated by topgen and then checked in so that they can be included with the `snippet` Hugo shortcode:
```
{{< snippet "../../../top_earlgrey/ip/pinmux/doc/autogen/targets.md" >}}
```

@moidx proposal to generate such tables with a build_docs pass is probably cleaner, so it probably makes sense to move the functions in `topgen/gen_top_docs.py` over to this file and consolidate the functions so that we can build both a summary pinmux table and a more detailed one.

Reply:
SG. I will defer the refactoring to a follow up PR.